README.md
# Mockery

[url-mock-2]:http://www.voidspace.org.uk/python/mock/
[url-ivan]:http://grishaev.me/
[url-foord]:http://www.voidspace.org.uk/python/weblog/index.shtml
[url-redefs]:https://clojuredocs.org/clojure.core/with-redefs-fn
[url-sling]:https://github.com/scgilardi/slingshot
[url-docs]:http://grishaev.me/mockery/
[url-tests]:https://github.com/igrishaev/mockery/blob/master/test/mockery/core_test.clj

Mockery is a simple and lightweight library to mock Clojure functions. It is
inspired by Python's [mock tool][url-mock-2] initially written
by [Michael Foord][url-foord].

## Installation

Add it to your dependencies:

`[mockery "0.1.4"]`

Mockery is used mostly for unit tests so it's better to put it into the `:test`
project's profile but not the global one.

## Why?

Imagine you have a function that fetches data from any remote service, say
Google or Twitter. On your dev machine, you don't have valid credentials or even
cannot access a service due to network configuration.

How to write tests for such code? Moreover, how to ensure you application
handles an incorrect HTTP response?

```
HTTP/1.1 403 OK

{"error": "wrong credentials"}
```

Or even non-JSON data (standard Nginx page):

```
HTTP/1.1 504 Gateway Timeout

<html><body><p>Gateway Timeout</p></body></html>
```

Will your application deal with such a response properly without crushing with
500 error? If you think it will, how could you guarantee that?

Now imagine you have modern micro-service architecture where each action
requires collecting data across 3 internal web-services making API calls and to
compose a final result.

Mockery helps to imitate such unusual behaviour and write unit tests that cover
all the cases were mentioned.

## Example

Mockery provides a `with-mock` macros. It substitutes a function with a
temporary one while the code block is being executed. Inside the code block, you
may call that function without performing unwanted I/O.

Under the hood, it relies on the standard
Clojure [clojure.core/with-redefs-fn][url-redefs] function. The changes are
visible across all the threads while macros works.

Quick example:

```clojure
(with-mock mock
  {:target ::test-fn ;; your function to mock
   :return 42}       ;; the result
  (test-fn 1 2 3))   ;; will return 42
```

The first parameter, `mock` in our example, is a variable name to bind a mock
object to. The second one is a map of options. The `:target` is the only
required key that should be a full-qualified keyword or symbol. For example,
`:clojure.string/join` or `'clojure.string/join`. If you'd like to point a
function located at the same namespace, use double colon syntax as well:
`::my-func` (expands into something like `:your.current.ns/my-func`).

## Features

Mockery works with the built-tin unit test framework as well:

```clojure
(defn mock-google [f]
  (with-mock _
    {:target :some.project/get-geo-point ;; makes API call to Google
     :return {:lat 14.2345235
              :lng 52.523513}} ;; what your function should return instead
    (f)))

(use-fixtures
  :each
  mock-google)

(deftest ...)
```

During the test, every `(get-geo-point ...)` call will return just the data you
specified without network communication.

`:return` value could be a function that will be called to produce further
result.


```clojure
(with-mock _
  {:target :myapp.google/get-geo-point
   :return (fn [& _] 100500)}
  (myapp.google/get-geo-point)) ;; returns 100500
```

The library imitates throwing bith Java and [Slingshot][url-sling] exceptions:

```clojure

(with-mock mock
  {:target ::test-fn
   :throw (Exception. "boom")}
  (try
    (test-fn 1)
    (catch Exception e
      (println e)))) ;; it was thrown

(with-mock mock
  {:target ::test-fn
   :throw {:type :domain/error ;; slingshot map
           :data "boom"}}
  (try+
   (test-fn 1)
   (catch [:type :domain/error] data
     (println data)))) ;; process data map here
```

Add side effects (prints, logs) passing a `:side-effect` key with a function
without arguments:

```clojure
(with-mock mock
   {:target ::my-func
    :return 42
    :side-effect #(println "Hello!")}
   (my-func 1))
;; in addition to the result, "Hello!" will appear.
```

Mockery helps to ensure the function was called the exact number of times with
proper arguments. The `mock` atom holds a state that changes while you call the
mocked function. It accumulates its arguments and the number of calls:

```clojure
(with-mock mock
  {:target ::test-fn
   :return 42}
  (test-fn 1)
  (test-fn 1 2)
  (test-fn 1 2 3)
  @mock)

;; returns (some fields are skipped)

{:called? true
 :call-count 3
 :call-args '(1 2 3)                  ;; the last args
 :call-args-list '[(1) (1 2) (1 2 3)] ;; args history

 :return 42                           ;; the last result
 :return-list [42 42 42]              ;; the entire result history
}
 ```

You may mock multiple functions at once using `with-mocks` macro:

```clojure
(with-mocks
 [foo {:target ::test-fn}
  bar {:target ::test-fn-2}]
 (test-fn 1)
 (test-fn-2 1 2)
 (is (= @foo
        {:called? true
         :call-count 1
         :call-args '(1)
         :call-args-list '[(1)]
         :target ::test-fn}))
 (is (= @bar
        {:called? true
         :call-count 1
         :call-args '(1 2)
         :call-args-list '[(1 2)]
         :target ::test-fn-2})))
```

## Other

For further reading, check out [Mockery documentation][url-docs]
and [unit tests][url-tests]. Feel free to submit your issues/suggestions.

Mockery was written by [Ivan Grishaev][url-ivan], 2018.


CHANGELOG.md
## 0.1.4
 * mock args added

## 0.1.3
 * `:return-list` field added

## 0.1.2
 * the main macros is rewritten completely

## 0.1.1-SNAPSHOT
 * slingshot dep removed

## 0.1.1
 * private variable bug fixed

## 0.1.0
 * Initial release


test/mockery/core_test.clj
(ns mockery.core-test
  (:require [clojure.test :refer :all]
            [slingshot.slingshot :refer [try+]]
            [mockery.core :refer [with-mock with-mocks]]))

(defn test-fn [a b]
  (+ a b))

(defn test-fn-2 [x y]
  (* x y))

(deftest test-no-calls
  (with-mock mock
    {:target ::test-fn}
    (is (= (-> @mock :called?)
           false))))

(deftest test-opt-as-var
  (let [opt {:target ::test-fn}]
    (with-mock mock opt
      (test-fn 1 2)
      (is (= (-> @mock :called?)
             true)))))

(deftest test-old-value
  (let [old1 (test-fn 1 2)]
    (with-mock mock
      {:target ::test-fn
       :return 42}
      (let [new1 (test-fn 1 2)]
        (is (= new1 42))))
    (let [old2 (test-fn 1 2)]
      (is (= old1 old2 3)))))

(deftest test-target-var
  (with-mock mock
    {:target ::test-fn
     :return 42}
    (let [result (test-fn 1 2)]
      (is (= result 42))
      (is (= (-> @mock (dissoc :target))
             {:called? true
              :call-count 1
              :call-args '(1 2)
              :call-args-list '[(1 2)]
              :return-list [42]
              :return 42})))))

(deftest test-multiple-calls
  (with-mock mock
    {:target ::test-fn
     :return 42}
    (test-fn 1)
    (test-fn 1 2)
    (test-fn 1 2 3)
    (is (= (-> @mock (dissoc :target))
           {:called? true
            :call-count 3
            :call-args '(1 2 3)
            :call-args-list '[(1) (1 2) (1 2 3)]
            :return 42
            :return-list [42 42 42]}))))

(deftest test-target-symbol
  (with-mock mock
    {:target 'mockery.core-test/test-fn
     :return 42}
    (test-fn 1)
    (is (= (-> @mock :call-count)
           1))))

(deftest test-throw
  (with-mock mock
    {:target ::test-fn
     :throw (Exception. "boom")}
    (is (thrown? Exception
                 (test-fn 1))) ))

(deftest test-slingshot
  (with-mock mock
    {:target ::test-fn
     :throw {:type :boom}}
    (try+
     (test-fn 1)
     (is false "should not be executed")
     (catch [:type :boom] _
       (is true)))))

(deftest test-side-effect
  (let [trap (atom false)
        effect #(reset! trap true)]
    (with-mock mock
      {:target ::test-fn
       :side-effect effect}
      (test-fn 1)
      (is @trap))))

(deftest test-return-fn
  (with-mock mock
    {:target ::test-fn
     :return #(+ 1 2 3)}
    (is (= (test-fn) 6))))

(deftest test-mock-multiple
  (with-mocks
    [foo {:target ::test-fn}
     bar {:target ::test-fn-2}]
    (test-fn 1)
    (test-fn-2 1 2)
    (is (= @foo
           {:called? true
            :call-count 1
            :call-args '(1)
            :call-args-list '[(1)]
            :target ::test-fn
            :return-list [nil]}))
    (is (= @bar
           {:called? true
            :call-count 1
            :call-args '(1 2)
            :call-args-list '[(1 2)]
            :target ::test-fn-2
            :return-list [nil]}))))

(deftest test-various-results
  (let [state (atom 0)
        func (fn [& _]
               (swap! state inc)
               @state)]
    (with-mocks
      [foo {:target ::test-fn
            :return func}]

      (test-fn :a)
      (test-fn :b)
      (test-fn :c)

      (is (= (dissoc @foo :return)
             {:called? true
              :call-count 3
              :call-args '(:c)
              :call-args-list '[(:a) (:b) (:c)]
              :return-list [1 2 3]
              :target ::test-fn})))))


test/mockery/foreign_ns.clj
(ns mockery.foreign-ns)

(defn foreign-fn
  [x y]
  (* x y))


src/mockery/core.clj
(ns mockery.core
  "A library to mock Clojure functions.")

(def ^:private
  +mock-defaults+
  "Mock default fields."
  {:called? false
   :call-count 0
   :call-args nil
   :call-args-list []
   :return-list []})

(defn- keyword-to-symbol
  "Turns a keyword into a symbold."
  [kwd]
  (apply symbol ((juxt namespace name) kwd)))

(defn make-mock
  "Retuns an atom the represents a mock."
  [opt]
  (-> +mock-defaults+
      (merge opt)
      atom))

(defn- smart-throw
  [exc]
  (cond
    (instance? Exception exc)
    (throw exc)

    (map? exc)
    (throw (ex-info "Mockery exception" exc))

    :default
    (assert false (format "Unsupported exception data: %s" exc))))

(defn eval-mock [mock args]
  (let [return (:return @mock)]
    (if (fn? return)
      (apply return args)
      return)))

(defn make-mock-fn
  "Returns a function to substitute a target.
  Calling this function will change the mock's state."
  [mock]
  (fn [& args]

    ;; update mock fields
    (swap! mock assoc :called? true)
    (swap! mock update :call-count inc)
    (swap! mock assoc :call-args args)
    (swap! mock update :call-args-list conj args)

    ;; trigger side effect
    (when-let [side-effect (:side-effect @mock)]
      (side-effect))

    ;; trigger exception
    (when-let [exc (:throw @mock)]
      (smart-throw exc))

    ;; eval the value, save it and return
    (let [result (eval-mock mock args)]
      (swap! mock update :return-list conj result)
      result)))

(defn- qualified?
  [symbol]
  (not (nil? (namespace symbol))))

(defn resolve-failure [target]
  (assert false (format "Cannot resolve target: %s" target)))

(defn resolve!
  "Resolves a variable by either a symbol or keyword.
  Rises an assertion error if not found."
  [target]
  (cond

    (symbol? target)
    (if (qualified? target)
      (if-let [ref (resolve target)]
        ref
        (resolve-failure target))
      (resolve-failure target))

    (keyword? target)
    (recur (keyword-to-symbol target))

    :default
    (resolve-failure target)))

(defmacro with-mocks
  "Like `with-mock` but for multiple mocks at once. `bind-opt` is a
  vector where each odd item is a bind symbol and even item is an
  option map. See `with-mock` for detailed description of each
  parameter."
  [bind-opt & body]
  `(let [~@(for [[i# el#] (map-indexed vector bind-opt)]
             (if (even? i#)
               el#
               `(make-mock ~el#)))]
     (with-redefs-fn
       (hash-map
        ~@(for [[i# el#] (map-indexed vector (reverse bind-opt))]
            (if (even? i#)
              `(resolve! (:target ~el#))
              `(make-mock-fn ~el#))))
       (fn []
         ~@body))))

(defmacro with-mock
  "Runs the code block substituting a target function with a dummy
  one (\"mock\" function). This dummy function behaves the way you
  determine with options. It counts the number of calls, accumulates
  it's arguments, can cause side effects or rise a given
  exception. Once you go out from the macro, the target function will
  obtain it's origin value.

  Arguments:

  - `mock`: a symbol to bind a mock instance inside the code block,

  - `opt`: a map with the following parameters:

  -- `:target` (required): a symbol or a keyword points to a target
  function to mock. May include a namespace (be full-qualified).

  -- `:value`: any value to return from a mocked function. If it's a
  function by itself (defn, fn or #(...)), it will be called
  afterwards.

  -- `:side-effect`: any function with no arguments that is triggered
  when calling a target function.

  -- `:throw`: either an Exception instance to be thrown with the
  standard `throw` pipeline or any map to be thrown with `(ex-info)`
  wrapper. Useful to simulate runtime exceptions.

  More on mock instance.

  The mock instance you have bound to the `mock` parameter is an atom
  holds a map. It extends the `opt` map described above. In addition
  to those immutable fields, it has the following ones:

  -- `:called?`: a boolean flag indicates whether a function was called
  at least one time or not. `false` by default.

  -- `:call-count`: a number times the function was called. 0 by default.

  -- `:call-args`: the last arguments were passed to the
  function. `nil` by default.

  -- `:call-args-list`: a vector of all the args were passed, `[]` by
  default.

  -- `:return`: the latest returned value.

  -- `:return-list`: a vector of all the values have ever been returned.

  Example:

  (with-mock mock
    {:target :clojure.pprint/pprint
     :return \":-)\"
     :side-effect #(println \"Hi!\")}
    (clojure.pprint/pprint {:foo 42}))

  Hi!
  :-)

  Once you deref the `mock` variable inside the code block, you'll get:

  {:called? true
   :call-count 1
   :call-args ({:foo 42})
   :call-args-list [({:foo 42})]
   :target :clojure.pprint/pprint
   :return \":-)\"
   :side-effect #function[mockery.core/eval12666/fn--12667]}

  Checking `:call-count` and `:call-args` might be good for your unit
  tests."
  [mock opt & body]
  `(with-mocks [~mock ~opt]
     ~@body))


index.clj
{:namespaces
 ({:doc "A library to mock Clojure functions.",
   :name "mockery.core",
   :wiki-url "https://igrishaev.github.io/mockeryindex.html",
   :source-url
   "https://github.com/igrishaev/mockery/blob/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj"}),
 :vars
 ({:raw-source-url
   "https://github.com/igrishaev/mockery/raw/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj",
   :name "make-mock",
   :file "src/mockery/core.clj",
   :source-url
   "https://github.com/igrishaev/mockery/blob/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj#L18",
   :line 18,
   :var-type "function",
   :arglists ([opt]),
   :doc "Retuns an atom the represents a mock.",
   :namespace "mockery.core",
   :wiki-url
   "https://igrishaev.github.io/mockery/index.html#mockery.core/make-mock"}
  {:raw-source-url
   "https://github.com/igrishaev/mockery/raw/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj",
   :name "make-mock-fn",
   :file "src/mockery/core.clj",
   :source-url
   "https://github.com/igrishaev/mockery/blob/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj#L43",
   :line 43,
   :var-type "function",
   :arglists ([mock]),
   :doc
   "Returns a function to substitute a target.\nCalling this function will change the mock's state.",
   :namespace "mockery.core",
   :wiki-url
   "https://igrishaev.github.io/mockery/index.html#mockery.core/make-mock-fn"}
  {:raw-source-url
   "https://github.com/igrishaev/mockery/raw/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj",
   :name "resolve!",
   :file "src/mockery/core.clj",
   :source-url
   "https://github.com/igrishaev/mockery/blob/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj#L75",
   :line 75,
   :var-type "function",
   :arglists ([target]),
   :doc
   "Resolves a variable by either a symbol or keyword.\nRises an assertion error if not found.",
   :namespace "mockery.core",
   :wiki-url
   "https://igrishaev.github.io/mockery/index.html#mockery.core/resolve!"}
  {:raw-source-url
   "https://github.com/igrishaev/mockery/raw/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj",
   :name "with-mock",
   :file "src/mockery/core.clj",
   :source-url
   "https://github.com/igrishaev/mockery/blob/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj#L113",
   :line 113,
   :var-type "macro",
   :arglists ([mock opt & body]),
   :doc
   "Runs the code block substituting a target function with a dummy\none (\"mock\" function). This dummy function behaves the way you\ndetermine with options. It counts the number of calls, accumulates\nit's arguments, can cause side effects or rise a given\nexception. Once you go out from the macro, the target function will\nobtain it's origin value.\n\nArguments:\n\n- `mock`: a symbol to bind a mock instance inside the code block,\n\n- `opt`: a map with the following parameters:\n\n-- `:target` (required): a symbol or a keyword points to a target\nfunction to mock. May include a namespace (be full-qualified).\n\n-- `:value`: any value to return from a mocked function. If it's a\nfunction by itself (defn, fn or #(...)), it will be called\nafterwards.\n\n-- `:side-effect`: any function with no arguments that is triggered\nwhen calling a target function.\n\n-- `:throw`: either an Exception instance to be thrown with the\nstandard `throw` pipeline or any map to be thrown with `(ex-info)`\nwrapper. Useful to simulate runtime exceptions.\n\nMore on mock instance.\n\nThe mock instance you have bound to the `mock` parameter is an atom\nholds a map. It extends the `opt` map described above. In addition\nto those immutable fields, it has the following ones:\n\n-- `:called?`: a boolean flag indicates whether a function was called\nat least one time or not. `false` by default.\n\n-- `:call-count`: a number times the function was called. 0 by default.\n\n-- `:call-args`: the last arguments were passed to the\nfunction. `nil` by default.\n\n-- `:call-args-list`: a vector of all the args were passed, `[]` by\ndefault.\n\n-- `:return`: the latest returned value.\n\n-- `:return-list`: a vector of all the values have ever been returned.\n\nExample:\n\n(with-mock mock\n  {:target :clojure.pprint/pprint\n   :return \":-)\"\n   :side-effect #(println \"Hi!\")}\n  (clojure.pprint/pprint {:foo 42}))\n\nHi!\n:-)\n\nOnce you deref the `mock` variable inside the code block, you'll get:\n\n{:called? true\n :call-count 1\n :call-args ({:foo 42})\n :call-args-list [({:foo 42})]\n :target :clojure.pprint/pprint\n :return \":-)\"\n :side-effect #function[mockery.core/eval12666/fn--12667]}\n\nChecking `:call-count` and `:call-args` might be good for your unit\ntests.",
   :namespace "mockery.core",
   :wiki-url
   "https://igrishaev.github.io/mockery/index.html#mockery.core/with-mock"}
  {:raw-source-url
   "https://github.com/igrishaev/mockery/raw/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj",
   :name "with-mocks",
   :file "src/mockery/core.clj",
   :source-url
   "https://github.com/igrishaev/mockery/blob/987480de78819062fd07da8ef09360812c408401/src/mockery/core.clj#L94",
   :line 94,
   :var-type "macro",
   :arglists ([bind-opt & body]),
   :doc
   "Like `with-mock` but for multiple mocks at once. `bind-opt` is a\nvector where each odd item is a bind symbol and even item is an\noption map. See `with-mock` for detailed description of each\nparameter.",
   :namespace "mockery.core",
   :wiki-url
   "https://igrishaev.github.io/mockery/index.html#mockery.core/with-mocks"})}
